python/metaparticle_pkg/metaparticle.py

```diff
@@ -0,0 +1,104 @@
+import functools
+import os
+import signal
+import sys
+from metaparticle_pkg.docker_builder import DockerBuilder
+from metaparticle_pkg.docker_runner import DockerRunner
+from metaparticle_pkg.metaparticle_runner import MetaparticleRunner
+
+def is_in_docker_container():
+    mp_in_container = os.getenv('METAPARTICLE_IN_CONTAINER', None)
+    if mp_in_container in ['true', '1']:
+        return True
+    elif mp_in_container in ['false', '0']:
+        return False
+
+    try:
+        with open('/proc/1/cgroup', 'r+t') as f:
+            lines = f.read().splitlines()
+            last_line = lines[-1]
+            if 'docker' in last_line:
+                return True
+            elif 'kubepods' in last_line:
+                return True
+            else:
+                return False
+
+    except IOError:
+        return False
+
+
+def write_dockerfile(name, exec_file):
+    with open('Dockerfile', 'w+t') as f:
+        f.write("""FROM python:3-alpine
+
+COPY ./ /{}/
+RUN pip install -r /{}/requirements.txt
+
+CMD python /{}/{}
+    """.format(name, name, name, exec_file))
+
+
+def select_builder(buildSpec):
+    if buildSpec == 'docker':
+        return DockerBuilder()
+    raise Exception('Unknown buildSpec {}'.format(buildSpec))
+
+
+def select_runner(runSpec):
+    if runSpec == 'docker':
+        return DockerRunner()
+    if runSpec == 'metaparticle':
+        return MetaparticleRunner()
+    raise Exception('Unknown buildSpec {}'.format(runSpec))
+
+def get_name(options):
+    try:
+        return options['name']
+    except KeyError:
+        pass
+
+    return os.path.basename(os.getcwd())
+
+
+def containerize(repository, options={}):
```

Comment:
In Python it’s more normal to add options as optional args to the function signature. It documents them nicely and helps linters. You can still add a catchall `**options` if there’s a use for it.

Reply:
Oh I forgot, giving a mutable default isn’t safe in Python, unlike Javascript. It won’t happen with this code, but that dict could change between calls and be a hard to trace bug in the future. Often `None` is used as the default, or an immutable version of the object like tuple instead of list. Unfortunately there’s no immutable dict though.